<?xml version="1.0"?>

<launch>
    <arg name="bag_path" default="" />
    <arg name="input_method" default=""/>
    <arg name="lidar_x_ext" default=""/>
    <arg name="lidar_y_ext" default=""/>   
    <arg name="lidar_z_ext" default=""/>
    
    <group if="$(eval input_method == 'rosbag')">
        <remap from="/kitti/velo/pointcloud" to="/cloud"/>
        <remap from="/hesai/pandar" to="/cloud"/>
        <param name="/use_sim_time" value="true" />
        <node pkg="rosbag" type="play" name="play" args="$(arg bag_path) --delay=5 --clock" />
    </group>

    <group if="$(eval input_method == 'sensor')">
        <remap from="/hesai/pandar" to="/cloud"/>
        <include file="$(find hesai_lidar)/launch/hesai_lidar.launch">
        <arg name="lidar_type" value="Pandar64"/>
        <arg name="frame_id" value="velo_link"/>
        </include>
        <param name="/use_sim_time" value="false" />
    </group>

    <node pkg="tf2_ros" type="static_transform_publisher" name="link2_broadcaster" args="0 0 0 0 0 0  base_link base_footprint" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="base_lidar_publisher" args="$(arg lidar_x_ext) $(arg lidar_y_ext) $(arg lidar_z_ext) 0 0 0  base_link velo_link" />
    

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(env HOME)/park_slam/src/rviz/slam_toolbox_default.rviz" />

    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server"/>
    <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
        <remap from="cloud_in" to="/cloud" />
        <param name="max_height" value="3.0" />
        <param name="min_height" value="-1.5" />
        <param name="scan_time" value="0.05" />
        <param name="range_min" value="3.5"/>
        <param name="range_max" value="35"/>
        <param name="use_inf" value="false"/>
    </node>


    <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher">
        <param name="use_odom" value="false" />
        <param name="use_imu" value="false" />
        <param name="use_cloud_input" value="false" />
        <param name="fixed_frame" value="odom" />
        <param name="restart" value="1"/>
        <param name="use_alpha_beta" value="true"/>
        <param name="max_correspondence_dist" value="1.0"/>
        <param name="max_linear_correction" value="1.0"/>
    </node>

    <node pkg="slam_toolbox" type="sync_slam_toolbox_node" name="slam_toolbox" output="screen">
        <rosparam command="load" file="$(env HOME)/park_slam/src/launch/slam_toolbox_params.yaml" />
        <param name="debug_logging" value="false"/>
    </node>


</launch>
